src/web/releases.rs

```diff
@@ -244,57 +244,94 @@ fn get_releases_by_owner(
+                crates.github_stars,
+                SUM(releases.downloads) AS downloads,
+
+                -- The levenshtein distance between the search query and the crate's name
+                levenshtein_less_equal($1, crates.name, 3) as distance,
+                -- The similarity of the tokens of the search vs the tokens of `crates.content`.
+                -- The `32` normalizes the number by using `rank / (rank + 1)`
+                ts_rank_cd(crates.content, to_tsquery($2), 32)  as content_rank
+            FROM releases INNER JOIN crates on releases.crate_id = crates.id
+
+            -- Filter crates that haven't been built and crates that have been yanked
+            WHERE releases.rustdoc_status = true
+                AND releases.yanked = false
+                AND (
+                    -- Crates names that match the query sandwiched between wildcards will pass
+                    crates.name ILIKE CONCAT('%', $1, '%')
+                    -- Crate names with which the levenshtein distance is closer or equal to 3 will pass
+                    OR levenshtein_less_equal($1, crates.name, 3) <= 3
+                    -- Crates where their content matches the query will pass
+                    OR to_tsquery($2) @@ crates.content
+                )
+            GROUP BY crates.id
+            -- Ordering is prioritized by how closely the query matches the name, how closely the
+            -- query matches the description finally how many downloads the crate has
+            ORDER BY distance DESC,
+                content_rank DESC,
+                SUM(downloads) DESC
+            -- Allows pagination
+            LIMIT $3 OFFSET $4",
+            &[&query, &split_query, &limit, &offset],
+        )
+        .ok()?;
 
-    for row in &rows {
-        let package = Release {
+    let packages: Vec<Release> = rows
+        .into_iter()
+        .map(|row| Release {
             name: row.get(0),
             version: row.get(1),
             description: row.get(2),
             target_name: row.get(3),
             release_time: row.get(4),
             rustdoc_status: row.get(5),
-            ..Release::default()
-        };
-
-        packages.push(package);
-    }
+            stars: row.get(6),
+        })
+        .collect();
 
     if !packages.is_empty() {
-        // get count of total results
+        // Get the total number of results that the query matches
         let rows = conn
             .query(
-                "SELECT COUNT(*) FROM crates WHERE content @@ to_tsquery($1)",
-                &[&query],
+                "SELECT COUNT(*) FROM crates
+                WHERE crates.name ILIKE CONCAT('%', CAST($1 AS TEXT), '%')
+                    OR levenshtein_less_equal(CAST($1 AS TEXT), crates.name, 3) <= 3
+                    OR crates.content @@ to_tsquery(CAST($2 AS TEXT))",
+                &[&(query as &str), &(&split_query as &str)],
             )
             .unwrap();
```

Comment:
This doesn't look right - it's redoing the query, but it already knows how many rows there are: it's `packages.len()`. Can you use that instead?

Reply:
It's not, actually. This gets the *total* number of matching crates, not the number of returned crates. This is for pagination (eventually), e.g. `Showing 50 out of 500`

Reply:
Ok, in that case this should be updated to match the other query and take `rustdoc_status` and `yanked` into account. Maybe we could turn that into a postgres function?

Reply:
You can select `COUNT(*) OVER()` to get the total count ignoring limit in the first query.